source/extensions/filters/http/cache/config.cc

```diff
@@ -2,31 +2,56 @@
 
 #include "source/extensions/filters/http/cache/cache_filter.h"
 
+#include "http_cache.h"
+
 namespace Envoy {
 namespace Extensions {
 namespace HttpFilters {
 namespace Cache {
 
-Http::FilterFactoryCb CacheFilterFactory::createFilterFactoryFromProtoTyped(
-    const envoy::extensions::filters::http::cache::v3::CacheConfig& config,
-    const std::string& stats_prefix, Server::Configuration::FactoryContext& context) {
+namespace {
+template <typename FactoryCtx>
+HttpCacheSharedPtr getCache(const envoy::extensions::filters::http::cache::v3::CacheConfig& config,
+                            FactoryCtx& ctx) {
   const std::string type{TypeUtil::typeUrlToDescriptorFullName(config.typed_config().type_url())};
-  HttpCacheFactory* const http_cache_factory =
-      Registry::FactoryRegistry<HttpCacheFactory>::getFactoryByType(type);
-  if (http_cache_factory == nullptr) {
+  auto* http_cache_factory = Registry::FactoryRegistry<HttpCacheFactory>::getFactoryByType(type);
+  if (Registry::FactoryRegistry<HttpCacheFactory>::getFactoryByType(type) == nullptr) {
     throw EnvoyException(
         fmt::format("Didn't find a registered implementation for type: '{}'", type));
   }
 
-  auto cache = http_cache_factory->getCache(config, context);
+  auto cache = http_cache_factory->getCache(config, ctx);
+  if (cache == nullptr) {
+    throw EnvoyException(fmt::format("Didn't find HTTP cache for type: '{}', config: '{}'", type,
+                                     config.DebugString()));
+  }
+
+  return cache;
+}
+} // namespace
+
+Http::FilterFactoryCb CacheFilterFactory::createFilterFactoryFromProtoTyped(
+    const envoy::extensions::filters::http::cache::v3::CacheConfig& config,
+    const std::string& stats_prefix, Server::Configuration::FactoryContext& context) {
+
+  auto cache = getCache(config, context);
 
-  return [config, stats_prefix, &context,
-          cache](Http::FilterChainFactoryCallbacks& callbacks) -> void {
-    callbacks.addStreamFilter(std::make_shared<CacheFilter>(config, stats_prefix, context.scope(),
-                                                            context.timeSource(), *cache));
+  return [config, stats_prefix, cache,
+          &context](Http::FilterChainFactoryCallbacks& callbacks) -> void {
+    callbacks.addStreamFilter(std::make_shared<CacheFilter>(stats_prefix, context.scope(),
+                                                            context.timeSource(), config, cache));
   };
 }
 
+Router::RouteSpecificFilterConfigConstSharedPtr
+CacheFilterFactory::createRouteSpecificFilterConfigTyped(
+    const envoy::extensions::filters::http::cache::v3::CacheConfig& config,
```

Comment:
Is it safe for createFilterFactoryFromProtoTyped to store a reference to `config` into its return value? (I couldn't find any documented guarantees.)

Reply:
This overall pattern of holding onto refs in constructed objects is common in Envoy.

In this case I am not sure if it's safe though...configs get updated via xDS during operation and I'm not clear on the lifetime of a config.

Looking at the usage i think we are just copying a couple of simple fields though. Instead of storing the config by reference can we just copy it? Is it very large? If so we could just copy explicitly the fields that matter for the lookup.

Reply:
The config is captured by copy in the lambda (line 39). I believe the FilterFactory is guaranteed to persist as long as any filter using it. ["This happens after all workers have discarded the previous config so it can be safely deleted on the main thread by an update with the new config."](https://sourcegraph.com/github.com/envoyproxy/envoy@b88641b/-/blob/source/common/filter/config_discovery_impl.h?L134) (though it's not entirely unambiguous.)

Copying the proto (of unknown size because it has a type-extension field in it) would be pretty expensive as a thing to do per filter instance. Just taking the specific fields would be dubious because the whole point is that fields from the type-extension field might be needed by the cache implementation (though potentially those fields could be bound up through getCache into the cache instance shared_ptr).